Add SAMD5x/E5x and all missing SAMx7x devices

chris-durand opened the pull request:
- [x] Add SAMD51 / SAME5x device family
  - [x] UART
  - [x] Clock
  - [x] USB 
- [x] Add SAME54 Xplained Pro board with blink example
- [x] Rename UART drivers
  - [x] `sam` -> `sam-sercom`
  - [x] `samg` -> `sam`
- [x] Improve support for SAMx7x device family
  - [x] Add UART / USART
  - [x] Add SAMS70, SAME70, SAMV71 devices
- [x] Fix data reception in the SAMx7x and SAMG UART driver
- [x] Add SAMV71 Xplained Ultra board with blink example
- [x] Refactor GCLK driver
  - [x] Expose flexible clock generator / clock source configuration in API
  - [x] Remove hardcoded assumptions of available clock sources 
  - [x] DFLL48 configuration
  - [x] DPLL configuration with automatic coefficient calculation
  - [x] API to enable peripheral clocks
- [x] Fix selection of device headers. In some cases headers belonging to the wrong device variant were chosen.
- [x] Adapt family naming scheme
  Not all devices which belong to one logical family share a common name prefix (e.g. the E70/S70/V70/V71 device family). The identifier scheme is changed to
  ```
  {platform}{series}{pin}{flash}{variant}-{package}{grade}
  ```
  and families are assigned based on logical grouping and not by name prefix.
- [x] Hardware test
   - [x] SAMD21 Xplained Pro
      - [x] UART
      - [x] GCLK
         - [x] DFLL48 open-loop
         - [x] DFLL48 closed-loop
         - [x] DPLL96
         - [x] OSC8M
         - [x] Peripheral clock configuration
     - [x] USB
   - [x] SAME54 Xplained Pro
      - [x] UART
      - [x] GCLK
      - [x] USB
   - [x] SAMV71 Xplained Ultra

SAMx7x have both UART and USART peripherals with the same id. The driver is adapted to provide `:platform:uart:*` and `:platform:usart:*` modules. This driver is also shared with the SAMG devices. For implementation simplicity and consistency I have changed the SAMG USART peripheral to also use the `:platform:usart` module and naming (see [this commit](https://github.com/modm-io/modm/commit/a5716d3ff6b66f7eed1815add4a8fcdd5dc5233b)). This is a breaking change. Would that be fine with you @mcbridejc @salkinium ? 

I have also renamed the previous "samg" UART driver to "sam" because this one is used by all SAM devices without SERCOM peripheral. The SERCOM based driver driver is renamed from "sam" to "sam-sercom". That renaming is a bit too much for github to display correctly in the diff.

chris-durand committed:
[repo] Enable SAMD5x/SAME5x device family

chris-durand committed:
[sam] Add minimum GCLK support for D5x/E5x devices

chris-durand committed:
[board] Add SAME54 Xplained Pro board

chris-durand committed:
[example] Add SAME54 Xplained Pro blink example

chris-durand committed:
[board] Add SAMV71 Xplained Ultra

chris-durand committed:
[example] Add SAMV71 Xplained Ultra blink example

chris-durand committed:
[sam] Rename UART drivers

The current "samg" driver is very common among SAM devices and is
renamed to "sam". The SERCOM UART driver is renamed to "sam-sercom".

chris-durand committed:
[sam] Add SAMx7x U(S)ART support, fix data reception

SAMx7x devices feature both UART and USART peripherals with the same id.
The driver will provide `:platform:uart:*` modules for UART instances
and `:platform:usart:*` modules for USART instances.

Fixes receiving data by enabling the RX interrupt.

**WARNING**, this is a breaking change.

For the sake of consistency, also SAMG devices, which have had a USART
peripheral provided as `UartX` with module `:platform:uart:*`, will use
the USART naming.

Furthermore, the clock source has to be specified as `U(s)art*` in
`SystemClock` to allow to configure it to sources different to `Mclk`.

chris-durand committed:
[repo] Enable complete SAMx7x device family

chris-durand committed:
[ext] Update sam and modm-devices submodules

chris-durand commented:
@salkinium In a lot of places we are making the assumption that device families have unique device name prefixes. For SAM this is not true, e.g. `SAMD51` and `SAME51`, `SAME53`, `SAME54` are a device family as well as `SAME70`, `SAMS70`, `SAMV70` and `SAMV71`. This makes the `family` field of the device identifier a lot less useful in the driver implementations. Can we live with it for now?

chris-durand commented:
The auto-generated HAL matrix table makes not much sense anymore with these devices added. The shown device families are D21, E, G55, S, V70.

mcbridejc commented:
> This is a breaking change. Would that be fine for you? 

That's fine by me.

chris-durand committed:
[core] Fix SAMx5x and SAMx7x delay

chris-durand committed:
[sam] Fix SAME5x gpio config

salkinium commented on src/modm/board/same54_xplained_pro/module.lb:
```diff
@@ -0,0 +1,32 @@
+    env.outbasepath = "modm/src/modm/board"
+    env.copy('board.hpp')
+    env.copy(repopath("tools/openocd/modm/atmel_same54_xplained_pro.cfg"), "atmel_same54_xplained_pro.cfg")
+    env.collect(":build:openocd.source", "modm/src/modm/board/atmel_same54_xplained_pro.cfg")
```
There is an special openocd folder for this: [`modm/openocd`](https://github.com/modm-io/modm/blob/develop/tools/build_script_generator/module.lb#L188).
[See this for usage](https://github.com/modm-io/modm/blob/develop/src/modm/board/nucleo_g474re/module.lb#L39-L41)

salkinium commented on examples/same54_xplained_pro/blink/main.cpp:
```diff
@@ -0,0 +1,14 @@
+#include "modm/board.hpp"
```
Copyright

chris-durand committed:
[sam] Fix ADC gpio data for D/E5x and x7x devices

chris-durand committed:
[sam] Fix macro name collision in ITM driver

chris-durand committed:
[sam] Add API to enable peripheral APB clocks on devices with GCLK

- Parse synchronous APB clock enable information from device headers
- Add GCLK API to enable / disable bus clocks

This change applies to all SAM devices with GCLK peripheral.

chris-durand committed:
[sam] Adapt peripheral drivers to GCLK changes

chris-durand committed:
[sam] Add SAMD5x/E5x USB support

chris-durand committed:
[sam] Fix systick frequency on SAMD5x/E5x

chris-durand committed:
[example] Add SAME54 Xplained PRO USB example

chris-durand committed:
[sam] Enable PMC clock driver for all devices with PMC

chris-durand committed:
[sam] Fix inconsistent macros for SAME70 "B"-variant

chris-durand committed:
[board] Adapt SAMD21 boards to GCLK changes

chris-durand committed:
[sam] Generate GCLK sources and peripheral clocks from device data

chris-durand committed:
[sam] Implement more complete GCLK driver for D2x/D5x/E5x devices

Extended and refactored GCLK driver. All implicit assumption that should
be in user code were removed from the driver. Clock generators can be
freely configured from user code.

Features:
- Flexible generator configuration
- Clock sources and peripheral sinks from device data
- External crystal, external 32k crystal support
- Internal oscillator configuration
- Dfll48 open/closed loop configuration

chris-durand committed:
[sam] Add DPLL driver with automatic coefficient calculation

chris-durand committed:
[board] Fix compilation of samd-mini bsp when using USB

rleh commented on src/modm/platform/clock/sam/gclk_impl.hpp.in:
```diff
@@ -12,12 +13,47 @@
+struct PeripheralClock<Peripherals::{{name | capitalize}}>
+%% endif
+{
+%% if clocks | length == 1
+%% set clock = clocks[0]
+	static inline void enable() { {{clock[0]}}->{{clock[1]}}MASK.reg |= (1u << {{clock[2]}}); }
+	static inline void disable() { {{clock[0]}}->{{clock[1]}}MASK.reg &= ~(1u << {{clock[2]}}); }
+%% else
```
Special case for clocks==1 is not necessary here.

chris-durand committed:
[sam] Adapt to new family naming scheme

salkinium commented:
Do you already want review feedback at this point?

chris-durand commented:
> Do you already want review feedback at this point?

I would appreciate a first review. Keep in mind that I haven't updated the PR description yet. There is a lot more stuff in there than mentioned.

chris-durand commented:
Just FYI, the diff does not correctly show that [in this commit](https://github.com/modm-io/modm/pull/827/commits/76c09fd691c1885f5676987a72e54c68e1b39d19) the uart drivers have been renamed. It might look like entirely new code in the diff but it isn't.

chris-durand committed:
[ci] Split SAM hal compile all job

chris-durand commented:
I had a strange issue with the CI. I have tried to update the CI to Ubuntu 22.04 and suddenly the hal quick compile job for same70q21b-cn failed. I have only changed the `runs-on` parameter in the github actions file. That should not have changed the container that is used.

It looked like it selected the wrong device headers for the "a"-variant for the "b"-variant device. Those have different uart defines. I absolutely can't reproduce the issue locally running exactly the same build commands that were failing on the CI. Any ideas?

salkinium commented on README.md:
```diff
@@ -121,9 +121,10 @@ Please [discover modm's peripheral drivers for your specific device][discover].
-<th align="center">V70</th>
+<th align="center">D1x/D2x/DAx</th>
+<th align="center">D5x/E5x</th>
+<th align="center">E7x/S7x/V7x</th>
```
```suggestion
<th align="center">E7x</br>S7x</br>V7x</th>
```
Would make the column less wide and not push the AVRs so far away…

salkinium commented on src/modm/board/samv71_xplained_ultra/board.hpp:
```diff
@@ -0,0 +1,109 @@
+{
+	// 300MHz system clock generated by PLLA from internal Rc 12MHz clock
+	static constexpr uint32_t PllAMult = 25;
+	static constexpr uint32_t Frequency = 300'000'000;
```
```suggestion
	static constexpr uint32_t Frequency = 300_MHz;
```

salkinium commented on src/modm/platform/clock/sam/module.lb:
```diff
@@ -19,11 +20,42 @@ def prepare(module, options):
+    driver = env[":target"].get_driver("gclk")
+
+    if target.family not in ["d1x/d2x/dax", "d5x/e5x"]:
+        raise RuntimeError("Unsupported device")
```
Should this not be done by returning `False` from the prepare function?

chris-durand replied:
All devices which have a `gclk` peripheral in their device files are structurally very similar to what is implemented in this driver. It is quite unlikely that someone would write a completely different clock driver for them. The only people who should ever encounter that error are those adding new device families to modm. We could move the check to the prepare function or remove these lines. In the end the result would be the same. In order to do something useful with the new devices this driver would need to be adapted. This exception is just a reminder.

salkinium commented on src/modm/platform/uart/cortex/itm.cpp.in:
```diff
@@ -134,6 +134,12 @@ Itm::discardTransmitBuffer()
 }
 
+%% if target.platform == "sam"
+// Some SAM headers define a PORT macro
```
dafuq??? 🤦

salkinium commented on src/modm/platform/uart/sam/usart/module.lb:
```diff
@@ -0,0 +1,89 @@
+            NumericOption(
+                name="buffer.rx",
+                description="Size of receive buffer",
+                minimum=1, maximum=2 ** 16 - 2,
```
```suggestion
                minimum=1, maximum="64K-2",
```
and above too. (Recently added to lbuild)

Is the minimum really 1? Ie. is unbuffered operation not implemented?

chris-durand replied:
Yes, it's not implemented in that driver. The other SAM driver, the sercom UART, only implements unbuffered mode. UART driver refactoring was on my list initially but this PR got huge enough. I'll do it after this is merged. Also the SAMx7x USB and external crystal support.

salkinium commented on tools/scripts/generate_hal_matrix.py:
```diff
@@ -202,7 +203,11 @@ def hal_create_table(targets, platforms, common_table=False):
 <th align="left">Peripheral</th>
 {% for fam in families  -%}
+{% if fam[0] == "sam" -%}
+<th align="center">{{ fam[1] | upper | replace("X", "x") }}</th>
```
```suggestion
<th align="center">{{ fam[1] | upper | replace("X", "x") | replace("/", "</br>") }}</th>
```
or something like that

chris-durand replied:
Good idea, thanks!

salkinium reviewed (commented):
Pretty good as per usual!

chris-durand commented:
I am planning to do the final hardware testing beginning of next week with D21, E54, and V71 boards. If everything goes well I'd like to merge soon after if possible.

chris-durand committed:
[sam] Fix SERCOM UART pin remapping

chris-durand committed:
[sam] Fix SAMD1x/D2x/DAx input pull-up configuration

setInput(InputType) erroneously reset the type to floating after
enabling pull-ups.

chris-durand committed:
[board] Add SAMD21 Xplained Pro

chris-durand committed:
[example] Add SAMD21 Xplained Pro blink example

chris-durand committed:
[example] Add SAMD21 Xplained Pro usb serial example

chris-durand commented:
@salkinium After fixing a few things the hardware tests are now successful. You could have a look at the added commits. Afterwards I would squash the fixup commits.

salkinium commented on src/modm/board/same54_xplained_pro/board.hpp:
```diff
@@ -0,0 +1,117 @@
+
+#define MODM_BOARD_HAS_LOGGER
+
+/// @ingroup modm_board_same54_xplained_pro
```
The ingroup thing need to be adapted on all new BSPs, since it doesn't work on namespaces.

chris-durand replied:
Ok, so I do

```
namespace Board
{
/// @ingroup modm_board_same54_xplained_pro
/// @{
…
/// @}
}
```

instead?

salkinium replied:
It's not enough, see the other more complex BSPs. :(

chris-durand replied:
fixed

salkinium commented on src/modm/board/samv71_xplained_ultra/board.hpp:
```diff
@@ -0,0 +1,109 @@
+
+#define MODM_BOARD_HAS_LOGGER
+
+/// @ingroup modm_board_samv71_xplained_ultra
```
Same here

salkinium commented on src/modm/board/samv71_xplained_ultra/board.hpp:
```diff
@@ -0,0 +1,109 @@
+}
+
+/*
+// TODO: usb
```
Or remove

chris-durand replied:
USB will be added shortly after. I'd leave it for now. I just did not want to extend this huge PR any further.

salkinium reviewed (commented):
Nice, just some minor things.

salkinium reviewed (approved):
Thank you very much! (+ rebase please)

salkinium commented:
Note to self: https://github.com/modm-io/modm-devices/pull/72 needs to be merged first!

chris-durand committed:
[sam] Fix device header selection

In case a header version with a variant suffix and one without
exist it was not defined which one was selected.

chris-durand committed:
[sam] Do not use deprecated uart defines on x7x b-variant devices

salkinium reviewed (approved):
Thanks a lot! I will merge this at the latest tomorrow together with the modm-devices PR and update the Q3 release.

salkinium commented:
@rleh do you still want to review this PR?
